### PR TITLE
[IR] Improve debugging experience around destructuring declarations

### PR DESCRIPTION
**Pull request by kandersen**
This PR adjusts the line numbers around destructuring declarations for a better debugging experience.

In particular, superfluous steps have been removed, and in places the IR backend is aligned closer to that of the old backend.

The PR is split in 3 commits, so that one can review:
  - the old behavior, expressed by new, passing tests
  - the new/desired behavior, expressed by adjusting test expectations and muting accordingly
  - the changes to achieve the desired behavior.

The fixes to the failing lambda tests are incoming, and experimentation with the failing multi-line for-loops indicate that the issue is orthogonal to destructuring, and will not be pursued further here.

### CODE REVIEW

**Comment by udalov** [comment]
Some tests are failing, in particular `IrBlackBoxCodegenTestGenerated$MultiDecl`, `IrSourceRangesTestCaseGenerated` and `IrBlackBoxCodegenTestGenerated$JavaInterop`.

And also, project bootstrap doesn't succeed, here's the exception when compiling `:core:metadata.jvm:compileKotlin`:
```
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
File being compiled: /mnt/agent/work/17e640964edd2053/kotlin/core/metadata.jvm/src/org/jetbrains/kotlin/metadata/jvm/deserialization/JvmNameResolver.kt
    The root cause java.lang.IndexOutOfBoundsException was thrown at: org.jetbrains.kotlin.com.intellij.util.text.CharSequenceSubSequence.<init>(CharSequenceSubSequence.java:20)
    at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException(CodegenUtil.kt:238)
    <...>
Caused by: java.lang.IndexOutOfBoundsException: chars sequence.length:4735, start:2147483647, end:0
    at org.jetbrains.kotlin.com.intellij.util.text.CharSequenceSubSequence.<init>(CharSequenceSubSequence.java:20)
    at org.jetbrains.kotlin.com.intellij.util.text.ImmutableText.subSequence(ImmutableText.java:194)
    at org.jetbrains.kotlin.backend.jvm.lower.TypeOperatorLowering.visitTypeOperator(TypeOperatorLowering.kt:729)
    <...>
    at org.jetbrains.kotlin.backend.jvm.lower.TypeOperatorLowering.lower(TypeOperatorLowering.kt:63)
    at org.jetbrains.kotlin.backend.common.phaser.FileLoweringPhaseAdapter.invoke(PhaseBuilders.kt:120)
    at org.jetbrains.kotlin.backend.common.phaser.FileLoweringPhaseAdapter.invoke(PhaseBuilders.kt:116)
    at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
    at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invokeSequential(performByIrFile.kt:65)
    ... 41 more

```

**Comment by kandersen** [comment]
Huh, that's odd - it appears to be the same exception as the cause to all the failings.

Thanks for pointing this out, will have a look!

**Comment by kandersen** [comment]
I pushed a fix that resolves the CharSequence exception: TypeOperatorLowering was not identifying "no offsets" correctly.

This PR does change the output of `IrSourceRangesTestCaseGenerated`. Do we have any tests that actually exercises this output, rather than just assert _what_ the output is? In intellij-community perhaps?

**Comment by kandersen** [comment]
The error occured because this PR `UNDEFINED_OFFSETS`s the source info on the _subsequent_ loads of the temporary variable that holds the _thing being_ destructed. When that thing is from java we insert `IMPLICIT_NOTNULL` checks on every one of those gets.

If it's easy, I guess we could improve code quality and compiler complexity by not adding the check on subsequent loads of that temporary variable: by construction, the _first_ load _not_ failing the null-check would imply the subsequent null checks pass and therefore could be removed.

If it's not easy, I will just note that down as a TODO :)

**Comment by udalov** [comment]
> This PR does change the output of `IrSourceRangesTestCaseGenerated`. Do we have any tests that actually exercises this output, rather than just assert what the output is? In intellij-community perhaps?

No, but since these ranges are only used to generate line numbers, we can usually come up with a stepping test where behavior changes. For example, in this case, if you take the test `kt24258.kt`, change the function to:
```
fun testLazyNullString() {      // (1)
    val                         // (2)
        s: String =             // (3)
        lazyNullString          // (4)
}                               // (5)
```
and try to debug it, before your changes you'll get (4)->(2)->(5), and after the change it'll be (4)->(3)->(5).

I don't really know which behavior is better; IMHO in this case it doesn't matter at all, but maybe I'm missing something. Maybe this could be added as a test to `compiler/testData/debug/stepping`.

File: compiler/testData/debug/localVariables/destructuring/lambdaCustomComponentNsMultiline.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    operator fun component2(): String {
+        return "K"
+    }
+}
+
+fun foo(a: MyPair, block: (MyPair) -> String): String = block(a)
+
+fun box() {
+    foo(MyPair("X", "Y")) { (x, y) -> x + y }
```
Comment (udalov) [inline compiler/testData/debug/localVariables/destructuring/lambdaCustomComponentNsMultiline.kt:17]: Looks like you intended to make this test multiline but forgot? Right now it's equivalent to `lambdaCustomComponentNs.kt`

File: compiler/testData/ir/irText/expressions/destructuring1.kt.txt [hunk 2]
```diff
@@ -31,3 +31,4 @@ fun B.test() {
   val y: Int = (<this>, tmp0_container).component2()
   // }
 }
+
```
Comment (udalov) [inline compiler/testData/ir/irText/expressions/destructuring1.kt.txt:34]: Maybe such changes in irText tests should be reverted

**Comment by kandersen** [comment]
> before your changes you'll get (4)->(2)->(5), and after the change it'll be (4)->(3)->(5).

Yeah, that's a deliberately change.

It's a) in line with what the old backend does, and b) consistent with destructuring declarations where you get a line on each variable load as it's initialized and not on the `val` keyword.

File: compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/StatementGenerator.kt [hunk 3]
```diff
@@ -131,40 +135,54 @@ class StatementGenerator(
...
+        }
 
         return irBlock
     }
 
     fun declareComponentVariablesInBlock(
         multiDeclaration: KtDestructuringDeclaration,
         irBlock: IrStatementContainer,
-        containerValue: IntermediateValue
+        containerValueWithOffsets: (Boolean) -> IntermediateValue
```
Comment (udalov) [inline compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/StatementGenerator.kt:160]: I'd prefer to pass two `IntermediateValue`s here, something like `firstContainerValue: IntermediateValue, restContainerValue: IntermediateValue`.

File: compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/intermediate/RematerializableValue.kt [hunk 4]
```diff
@@ -31,9 +32,25 @@ fun Scope.createTemporaryVariableInBlock(
...
     block: IrStatementContainer,
-    nameHint: String? = null
+    nameHint: String? = null,
+    startOffset: Int = irExpression.startOffset,
+    endOffset: Int = irExpression.endOffset
 ): IntermediateValue {
-    val temporaryVariable = createTemporaryVariable(irExpression, nameHint)
+    val temporaryVariable = createTemporaryVariable(irExpression, nameHint, startOffset = startOffset, endOffset = endOffset)
     block.statements.add(temporaryVariable)
     return VariableLValue(context, temporaryVariable)
```
Comment (udalov) [inline compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/intermediate/RematerializableValue.kt:41]:
```suggestion
    return VariableLValue(context, declareTemporaryVariableInBlock(irExpression, block, nameHint, startOffset, endOffset))
```

**Comment by udalov** [comment]
> If it's easy, I guess we could improve code quality and compiler complexity by not adding the check on subsequent loads of that temporary variable: by construction, the first load not failing the null-check would imply the subsequent null checks pass and therefore could be removed.

Yeah, let's try doing this.

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt [hunk 5]
```diff
@@ -757,6 +757,9 @@ private class TypeOperatorLowering(private val backendContext: JvmBackendContext
         origin == IrDeclarationOrigin.DELEGATED_PROPERTY_ACCESSOR ||
                 origin == IrDeclarationOrigin.DELEGATED_MEMBER
 
+    private fun CharSequence.validSourcePosition(startOffset: Int, endOffset: Int): Boolean =
+        startOffset in 0 until endOffset && endOffset < length
```
Comment (udalov) [inline compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt:761]: I don't understand... The change is that now we skip the case when `startOffset >= endOffset`? Isn't it an incorrect state for non-undefined offsets to be in?

Comment (kandersen) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt:761]: It is! And that is the state that's returned by the `extents` function in the case that it only meets offsets < 0 in the children of the element. In that case it returns `MAX_INT to 0` and it fails this check.

We could perhaps replace that part of the logic with an Option or null or something, if you'd prefer to not complicate the arithmetics?

File: compiler/testData/debug/localVariables/destructuring/lambdaCustomComponentNsMultiline.kt (hunk 1 above)
Comment (kandersen) [reply compiler/testData/debug/localVariables/destructuring/lambdaCustomComponentNsMultiline.kt]: Thanks, absolutely right.

File: compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/StatementGenerator.kt (hunk 3 above)
Comment (kandersen) [reply compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/StatementGenerator.kt]: Done!

File: compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/intermediate/RematerializableValue.kt (hunk 4 above)
Comment (kandersen) [reply compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/intermediate/RematerializableValue.kt]: Nice, thanks.

**Comment by kandersen** [comment]
> Maybe this could be added as a test to `compiler/testData/debug/stepping`.

Oh, forgot to mention that it's indeed covered in `compiler/testData/debug/localVariables/assignment.kt` :)

File: compiler/testData/ir/irText/expressions/destructuring1.kt.txt (hunk 2 above)
Comment (kandersen) [reply compiler/testData/ir/irText/expressions/destructuring1.kt.txt]: (un)done!

**Comment by udalov** [comment]
There's also `FirLocalVariableTestGenerated` which is failing now.

Also, Kotlin/Native runtime compilation fails. To reproduce, put `kotlin.native.enabled=true` to `local.properties` and run `./gradlew dist`:
```
> Task :kotlin-native:runtime:linux_x64StdlibCache FAILED
e: Compilation failed: Assertion failed

 * Source files: 
 * Compiler version info: Konan: 1.6.0-dev / Kotlin: 1.7.255
 * Output kind: STATIC_CACHE

e: java.lang.AssertionError: Assertion failed
        at org.jetbrains.kotlin.backend.konan.llvm.DebugUtilsKt.location(DebugUtils.kt:99)
        at org.jetbrains.kotlin.backend.konan.llvm.DebugUtilsKt.line(DebugUtils.kt:108)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor$FunctionScope.location(IrToBitcode.kt:770)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor$InnerScope.location(IrToBitcode.kt)
        ...
```

**Comment by kandersen** [comment]
Thank you for repro instructions. 

Last I tried (admittedly 2+ years ago) building kotlin native wasn't as straight forward!

**Comment by kandersen** [comment]
> Last I tried (admittedly 2+ years ago) building kotlin native wasn't as straight forward!

Spoke a little too soon! :sweat_smile:  

```
CoverageMappingC.cpp:(.text+0x0): multiple definition of `LLVMWriteCoverageRegionMapping'; /usr/local/google/home/kjaa/kotlin/kotlin-native/llvmCoverageMappingC/build/libcoverageMapping.a(CoverageMappingC.o):CoverageMappingC.cpp:(.text+0x0): first defined here
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
        at org.jetbrains.kotlin.konan.exec.Command.handleExitCode(ExecuteCommand.kt:114)
        at org.jetbrains.kotlin.konan.exec.Command.getResult(ExecuteCommand.kt:103)
        at org.jetbrains.kotlin.konan.exec.Command.getOutputLines(ExecuteCommand.kt:82)
        at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.runCmd(main.kt:110)
        at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.processCLib(main.kt:383)
        at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.processCLibSafe(main.kt:214)
        at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.main(main.kt:68)

> Task :kotlin-native:backend.native:genLlvmInteropStubs FAILED
```

Also, re. the `FirLocalVariableTestGenerated` failures -- would a new `JVM_FIR` target backend make sense to facilitate muting? I don't think there's another mechanism in place.

**Comment by madsager** [comment]
> Also, re. the `FirLocalVariableTestGenerated` failures -- would a new `JVM_FIR` target backend make sense to facilitate muting? I don't think there's another mechanism in place.

Drive-by: does `IGNORE_BACKEND_FIR: JVM_IR` work here?

**Comment by kandersen** [comment]
> Drive-by: does `IGNORE_BACKEND_FIR: JVM_IR` work here?

Thanks for stopping by, @madsager, it absolutely does! 

I didn't leave enough room at the top for the directive so everything has to be pushed by a line, though :facepalm: :sweat_smile:

**Comment by kandersen** [comment]
Sorry for the noise, @udalov, I think my native build failed due to some stale caching or something to that effect.

1) I muted the FIR failures, they can be adressed in a later fix. Nothing _breaks_ out right, it just produces slightly different stepping behavior.

2) The native failure is an assertion that nothing has `UNDEFINED_OFFSET`. Replacing my uses of undefined with `SYNTHETIC_OFFSET` in this PR is all the same to JVM backend, but appeases native. Do you know of some debugging tests on native? Perhaps you know of who might be familiar with this?

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt (hunk 5 above)
Comment (udalov) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt:761]: Oh, looking at it again I'm not sure what confused me the first time. It's clear this way too. Thanks for the clarification!

**Comment by udalov** [comment]
> The native failure is an assertion that nothing has `UNDEFINED_OFFSET`. Replacing my uses of undefined with `SYNTHETIC_OFFSET` in this PR is all the same to JVM backend, but appeases native. Do you know of some debugging tests on native? Perhaps you know of who might be familiar with this?

@SvyatoslavScherbina Could you please help here, or suggest someone from K/N who can help?

**Comment by SvyatoslavScherbina** [comment]
> Do you know of some debugging tests on native?

See https://github.com/JetBrains/kotlin/tree/master/kotlin-native/backend.native/tests/debugger/src/test/kotlin/org/jetbrains/kotlin/native/test/debugger. E.g. `LldbTests.kt`.

Can be run with `:kotlin-native:backend.native:tests:debugger_test` task (enabled only on macOS at the moment).

**Comment by udalov** [comment]
@kandersen Please also mute `FirSteppingTestGenerated.testLocalProperty`.

**Comment by kandersen** [comment]
Done, thanks for catching that!

**Comment by udalov** [comment]
@kandersen Just in case, the changes are OK for me, but I was kind of waiting for resolution of https://github.com/JetBrains/kotlin/pull/4759#issuecomment-1062847215 -- I expect that either you can try to do this, or at least add a TODO and address it separately. :)

**Comment by udalov** [comment]
Now the (newly added) `JsIrLineNumberTestGenerated` is failing.

`testDestructuring`:
```
expected:
<1 1 * 1 1 * 9 9 9 9 9 6 6 9 7 7 9 11 11 12 12 15 15 * 26 26 * 17 17 16 18 18 16 20 20 21 21 * 1>
but was:
<1 1 * 1 1 * 9 9 9 9 9 6 6 1 9 7 7 1 11 11 12 12 15 15 * 26 26 * 17 17 1 16 18 18 1 16 20 20 21 21 * 1>
```

`testDestructuringInline`:
```
expected:
<1 1 * 1 1 * 9 9 9 9 9 * 31 31 9 6 6 * 33 33 9 7 7 11 11 12 12 15 15 * 26 26 * 29 29 29 29 * 29 29 * 29 29 * 31 31 * 33 33 * 31 31 16 17 17 * 33 33 16 18 18 20 20 21 21 * 1>
but was:
<1 1 * 1 1 * 9 9 9 9 9 * 31 31 9 6 6 1 * 33 33 1 7 7 1 11 11 12 12 15 15 * 26 26 * 29 29 29 29 * 29 29 * 29 29 * 31 31 * 33 33 * 31 31 16 17 17 1 * 33 33 16 18 18 1 20 20 21 21 * 1>
```

@broadwaylamb Would this change in test data be OK or should we investigate and change it somehow?

**Comment by broadwaylamb** [comment]
The generated JavaScript code with annotated line numbers for the [`destructuring.kt`](https://github.com/JetBrains/kotlin/blob/3b53d8694ff9e5a3532f5eff893a258b5a73ec21/js/js.translator/testData/lineNumbers/destructuring.kt) test is:

```js
var main = function (_, kotlin_kotlin) {
    'use strict';
    //region block: imports
    var Pair = kotlin_kotlin.$_$.i;
    var listOf = kotlin_kotlin.$_$.c;
    //endregion
    //region block: pre-declaration
    //endregion
    //region file: /destructuring.kt
    /* no-line */ function set_log(_set____db54di) {
      /* 1 1 */ log = _set____db54di;
    }
    /* no-line */ function get_log() {
      /* 1 1 */ return log;
    }
    var log;
    /* no-line */ function foo() {
      /* 9 9 */ var tmp0_iterator = listOf(new Pair('1', '2')).iterator_jk1svi_k$();
      /* 9 */ while (tmp0_iterator.hasNext_bitz1p_k$()) {
        /* 9 9 */ var tmp1_loop_parameter = tmp0_iterator.next_20eer_k$();
        /* 6 6 1 9 */ var q = tmp1_loop_parameter.component1_7eebsc_k$();
        /* 7 7 1 */ var w = tmp1_loop_parameter.component2_7eebsb_k$();
        /* 11 11 */ log = log + q;
        /* 12 12 */ log = log + w;
      }
      /* 15 15 */ bar(foo$lambda());
    }
    /* no-line */ function bar(f) {
      /* 26 26 */ f(new Pair('w', 'e'));
    }
    function foo$lambda() {
      /* no-line */ return function (_name_for_destructuring_parameter_0__wldtmu) {
        /* 17 17 1 16 */ var q = _name_for_destructuring_parameter_0__wldtmu.component1_7eebsc_k$();
        /* 18 18 1 16 */ var w = _name_for_destructuring_parameter_0__wldtmu.component2_7eebsb_k$();
        /* 20 20 */ log = log + q;
        /* 21 21 */ log = log + w;
      };
    }
    //endregion
    //region block: init
    /* 1 */ log = '';
    //endregion
    /* no-line */ return _;
  }(typeof main === 'undefined' ? {} : main, kotlin_kotlin);
```

What drags my attention is these lines:

```js
/* 6 6 1 9 */ var q = tmp1_loop_parameter.component1_7eebsc_k$();
/* 7 7 1 */ var w = tmp1_loop_parameter.component2_7eebsb_k$();
```

The numbers in the comments are the line numbers in the Kotlin file that these JS statements are translated from.

Note the 1s. This means that for some reason these statements contain source information that points to the first line of the original Kotlin file, which doesn't really make sense, since that line contains a declaration of the `log` variable.

Before this patch the generated code looked like this:

```js
/* 6 6 9 */ var q = tmp1_loop_parameter.component1_7eebsc_k$();
/* 7 7 9 */ var w = tmp1_loop_parameter.component2_7eebsb_k$();
```

**Comment by kandersen** [comment]
@broadwaylamb thank you providing the test output!

I am surprised by seeing the linenumber `1` in there. The patch in this PR changes the generated IR in psi2ir, and the source information is either composed from the PSI for the destructuring declarations or undefined, so my knee-jerk reaction is to think a later stage in the JS compilation may interpret UNDEFINED_OFFSET in a way that "rounds to" line 1. 

Do you have any ideas or other suspicions? I will be happy to have a closer look.

**Comment by broadwaylamb** [comment]
I will investigate.

**Comment by broadwaylamb** [comment]
I see.

`SYNTHETIC_OFFSET` is explicitly mapped to the zero line number (line numbers are zero-based in the internal representation) here:

https://github.com/JetBrains/kotlin/blob/476e65f93469c45e8c58ccfab89c81e1e73001f2/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/AdditionalIrUtils.kt#L208-L219

I'm not sure why that is, @udalov do you know anyone who could clarify?

**Comment by udalov** [comment]
I have no idea, to be honest. :/

Do you think it's a show-stopper for merging this change to master right now? My initial reaction was that it isn't, mainly because I noticed there are numerous such `1`s in other tests in `js/js.translator/testData/lineNumbers` and kind of thought that this support is not fully finished, so it'll need an investigation anyway. But if I'm wrong and you think that it's not OK to merge it as is, we can probably try to isolate the change so that it'll only apply for JVM.

**Comment by kandersen** [comment]
Thank you both for your considerations. I have rebased on master.

**Comment by kandersen** [comment]
@udalov Pushed a take on making the changes JVM specific. It feels a little clumsy, but is probably fine. What do you think? Would it be better to just "fork" `org.jetbrains.kotlin.psi2ir.generators.StatementGenerator#visitDestructuringDeclaration` at the top based on platform, at the cost of a bit of duplication?

Let me know how you want to proceed! :)

File: compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/StatementGenerator.kt [hunk 6]
```diff
@@ -110,8 +113,9 @@ class StatementGenerator(
             )
         }
 
+        val sourceElement = property.nameIdentifier ?: property
```
Comment (udalov) [inline compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/StatementGenerator.kt:116]:
```suggestion
        val sourceElement =
            if (context.extensions.debugInfoOnlyOnVariablesInDestructuringDeclarations) {
                property.nameIdentifier ?: property
            } else {
                property
            }
```
WIthout this change, Native debug tests (`kt42208 with variable`, `kt42208 with passing lambda to another function`) are still failing

**Review by udalov — CHANGES_REQUESTED** [review]

File: compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/GeneratorExtensions.kt [hunk 7]
```diff
@@ -48,4 +48,7 @@ open class GeneratorExtensions : StubGeneratorExtensions() {
 
     open val parametersAreAssignable: Boolean
         get() = false
+
+    open val debugInfoOnlyOnVariablesInDestructuringDeclarations: Boolean
```
Comment (udalov) [inline compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/GeneratorExtensions.kt:52]: Maybe add a little explanation what this does? (E.g. to simplify further implementation in FIR)

Comment (kandersen) [reply compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/GeneratorExtensions.kt:80]: Done, thank you!

**Comment by kandersen** [comment]
Revised and pushed.

**Review by udalov — APPROVED** [review]

**Comment by udalov** [comment]
Finally, the build succeeded. Merged, thank you!

**Pull request merged**